File: Typescript/ObserverPattern.js
"use strict";
// Subject (Observable)
class WeatherStation {
    constructor(location) {
        this.observers = [];
        this.temperature = null;
        this.humidity = null;
        this.location = location;
    }
    addObserver(observer) {
        if (!this.observers.includes(observer)) {
            this.observers.push(observer);
        }
    }
    removeObserver(observer) {
        const index = this.observers.indexOf(observer);
        if (index !== -1) {
            this.observers.splice(index, 1);
        }
    }
    notifyObserver() {
        for (const observer of this.observers) {
            observer.update(this.temperature, this.humidity, this.location, WeatherStation.name);
        }
    }
    setConditions(temperature, humidity) {
        this.temperature = temperature;
        this.humidity = humidity;
        this.notifyObserver();
    }
}
// ConcreteObserver
class WeatherDisplay {
    constructor(name) {
        this.name = name;
    }
    update(temperature, humidity, location, observerName) {
        console.log(`${this.name} received update from ${observerName}  at location ${location}- Temperature: ${temperature}°C, Humidity: ${humidity}%`);
    }
}
// Usage example
const station1 = new WeatherStation("Station 1");
const station2 = new WeatherStation("Station 2");
const display1 = new WeatherDisplay("Display 1");
const display2 = new WeatherDisplay("Display 2");
station1.addObserver(display1);
station1.addObserver(display2);
station2.addObserver(display1);
station2.addObserver(display2);
station1.setConditions(25, 60);
station2.setConditions(30, 50);
station1.removeObserver(display2);
station1.setConditions(28, 55);

File: Python/ObserverPattern.py
# Observer pattern implementation for a weather system

# Subject (Observable)
class WeatherStation:
    def __init__(self, location):
        self.location = location
        self._observers = []
        self._temperature = None
        self._humidity = None

    def add_observer(self, observer):
        if observer not in self._observers:
            self._observers.append(observer)

    def remove_observer(self, observer):
        self._observers.remove(observer)

    def set_conditions(self, temperature, humidity):
        self._temperature = temperature
        self._humidity = humidity
        self.notify_observers()

    def notify_observers(self):
        for observer in self._observers:
            observer.update(self._temperature, self._humidity, self.location)


# Observer (Observer)
class Display:
    def update(self, temperature, humidity, location):
        pass


# ConcreteObserver
class WeatherDisplay(Display):
    def update(self, temperature, humidity, location):
        print(f"Display at {location} updated - Temperature: {temperature}°C, Humidity: {humidity}%")


# Usage example
if __name__ == "__main__":
    # Create weather stations
    station1 = WeatherStation("Station 1")
    station2 = WeatherStation("Station 2")

    # Create displays
    display1 = WeatherDisplay()
    display2 = WeatherDisplay()

    # Register displays as observers for each station
    station1.add_observer(display1)
    station1.add_observer(display2)
    station2.add_observer(display1)
    station2.add_observer(display2)

    # Simulate weather changes
    station1.set_conditions(25, 60)
    station2.set_conditions(30, 50)

    # Unregister display2 from station1
    station1.remove_observer(display2)

    # Simulate another weather change
    station1.set_conditions(28, 55)
